### DISCUSSION TITLE
How to validate Definition using Go API

### DISCUSSION THREAD

**Discussion by heestree** [cue-lang/cue#2155] [Q&A]
Hi,

How can I mimic the behavior of `.\cue.exe vet --schema '#Def_2' cue-question.cue .\cue-question.json` with Go API?

In cue-question.cue
```
// .\cue.exe vet --schema '#Def_1' cue-question.cue .\cue-question.json
#Def_1: {
    Setup: {
        TestItem?: true
    }
}

// .\cue.exe vet --schema '#Def_2' cue-question.cue .\cue-question.json
#Def_2: {
    Setup: {
        TestItem: true
    }
}
```

In cue-question.json
```
{
    "Setup": {
        "TestItem": false
    }
}
```

I am using 
```
.\cue.exe vet --schema '#Def_1' cue-question.cue .\cue-question.json
```
or
```
.\cue.exe vet --schema '#Def_2' cue-question.cue .\cue-question.json
```

to validate #Def_1 or #Def_2 schema(or definition).
What I trying to do is doing the same thing with my Go codes by calling any Go API provided by Cue.
I know we can call validate() function in Go codes, but I don't know how to call validate or any other function to validate only #Def_1 or #Def_2 schema.
I would appreciate any help from cue experts.

**Comment by mvdan** [comment]
Here's a working example, which fails due to the `true & false` discrepancy: https://go.dev/play/p/b4S-VWK-yWR

In short, it uses `os.ReadFile` and `CompileBytes` to load both files, `LookupPath` to select a schema by a path, and `Unify` and `Validate` to mimic the validation logic of `cue vet`.

**Comment by rudifa** [comment]
@mvdan 

When I run your playground example above, I get this:
```
timeout running go build
go: downloading cuelang.org/go v0.5.0-beta.1
go: downloading github.com/cockroachdb/apd/v2 v2.0.2
go: downloading github.com/mpvl/unique v0.0.0-20150818121801-cbe035fff7de
go: downloading golang.org/x/text v0.3.8
go: downloading golang.org/x/net v0.0.0-20220722155237-a158d28d115b
go: downloading github.com/google/uuid v1.2.0
go: downloading github.com/pkg/errors v0.8.1
go: downloading gopkg.in/yaml.v3 v3.0.1

Go build failed.
```
What am I doing wrong?

**Reply by myitcv** [reply]
The Go playground doesn't, unfortunately, always work with respect to resolving third party modules. It's flakey in a rather frustratingly indeterminate way, in part related to https://github.com/golang/go/issues/50158

**Reply by rudifa** [reply]
Thank you for the clarification.

FWIW, this is the diagnostic in the browser, identical across Chrome, Brave, Safari, Firefox and Edge, on my Intel Mac; the timeout is about 10 seconds:

```
timeout running go build
go: downloading cuelang.org/go v0.5.0-beta.1
go: downloading github.com/cockroachdb/apd/v2 v2.0.2
go: downloading github.com/mpvl/unique v0.0.0-20150818121801-cbe035fff7de
go: downloading golang.org/x/text v0.3.8
go: downloading golang.org/x/net v0.0.0-20220722155237-a158d28d115b
go: downloading github.com/google/uuid v1.2.0
go: downloading github.com/pkg/errors v0.8.1
go: downloading gopkg.in/yaml.v3 v3.0.1

Go build failed.

```

**Reply by mvdan** [reply]
A narrower issue about this problem is one I filed a bit later: https://github.com/golang/go/issues/56977

**Reply by mvdan** [reply]
In fact I filed it precisely because of this thread :)
